### PR TITLE
Wordclock Web App

### PR DESCRIPTION
**Pull request by FrankX0**
I did some polishing-up of the App.
It now consists of two tabs: one for control, one for the colours.
Now fits better in the Wordclock structure of files.

Have fun!

### CODE REVIEW

File: wordclock_interfaces/templates/app.html [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            <v-content>
+                <v-container fluid fill-height>
+                    <v-layout align-center justify-center>
+                        <v-flex xs12 sm8 md4>
+                            <v-tabs dark color="primary" class="elevation-12">
+                                <v-tab ripple height="100%">
+                                    Control
+                                </v-tab>
+                                <v-tab ripple>
+                                    Colours
```
Comment (bk1285) [inline wordclock_interfaces/templates/app.html]: I'd prefer "Colors"

File: wordclock_interfaces/templates/app.html [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                                </v-tab>
```
Comment (bk1285) [inline wordclock_interfaces/templates/app.html]: How about adding another tab with some links to the project-page and API. In this case, the app would become ready to replace the current index.html... ;)

Comment (bk1285) [reply wordclock_interfaces/templates/app.html]: Links would be:
* The project page: https://github.com/bk1285/rpi_wordclock
* The API: wordclock-ip//api

File: wordclock_interfaces/templates/app.html [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                                                    </v-flex>
+                                                    <v-flex xs6>
+                                                        <v-card tile flat>
+                                                            <v-card-text>
+                                                                <v-btn color="primary" v-on:click="buttonClick('right')">Right</v-btn>
+                                                            </v-card-text>
+                                                        </v-card>
+                                                    </v-flex>
+                                                </v-layout>
+                                                <v-layout row justify-center>
```
Comment (bk1285) [inline wordclock_interfaces/templates/app.html]: Is there any chance to place all three buttons in one single row?

**Comment by bk1285** [comment]
Hey @FrankX0,
I just scrolled through the updates and did some tests with my clock. Very nice face lift! I like it a lot! :+1: 

Best,
Bernd

File: wordclock_interfaces/templates/app.html (hunk 2 above)
Comment (FrankX0) [reply wordclock_interfaces/templates/app.html]: Working on this one....

File: wordclock_interfaces/templates/app.html (hunk 1 above)
Comment (FrankX0) [reply wordclock_interfaces/templates/app.html]: Done

File: wordclock_interfaces/templates/app.html (hunk 3 above)
Comment (FrankX0) [reply wordclock_interfaces/templates/app.html]: Done

**Comment by bk1285** [comment]
Hey @FrankX0, the changes so far look good to me! API 4.2 is now on develop, allowing you to request the current color + brightness from the clock.
... in case you want to initialize the values.. ;)

File: wordclock_interfaces/templates/app.html (hunk 2 above)
Comment (FrankX0) [reply wordclock_interfaces/templates/app.html]: About page added.
See if you can find it :-)

**Comment by bk1285** [comment]
FYI: Added icons, which can be included as done here https://github.com/bk1285/rpi_wordclock/commit/0495dbd031d65303812da012bac38c61b7615f35

;)

**Comment by FrankX0** [comment]
I was also just working on this!
Please add the Apple icons: this will give a nice icon on the home screen.

File: wordclock_interfaces/templates/app.html (hunk 2 above)
Comment (bk1285) [reply wordclock_interfaces/templates/app.html]: Found it! Nice! :+1:

**Comment by bk1285** [comment]
Hi @FrankX0,

concerning the apple-icons: You refer to the resolution?

**Comment by FrankX0** [comment]
Not sure what you mean.
In my understanding for Apple products the icons should be available in different resolutions and names as indicated in the header:
- apple-icon-57x57.png
- apple-icon-60x60.png
- apple-icon-72x72.png
- apple-icon-76x76.png
- apple-icon-114x114.png
- apple-icon-120x120.png
- apple-icon-144x144.png
- apple-icon-152x152.png
- apple-icon-180x180.png

**Comment by bk1285** [comment]
Hey @FrankX0, 

just added the "requested" icons on develop.

Best,
Bernd

**Comment by bk1285** [comment]
Another point: Throughout the last years, I experienced a traffic peak on the git-repo within December: Lot's of people building this clock seem to have a hard deadline on December 24th.... :D

As mentioned before, I think your work advances the usability of the web-interface (compared to the current one) in a tremendous way! Therefore, I'd propose to keep the PR open until mid of November, in case you have any updates/fixes on this. Afterwards I'd merge into develop + master. Fine for you, @FrankX0 ?

**Comment by FrankX0** [comment]
No problem!
Maybe we can squeeze in API 4.2

**Comment by FrankX0** [comment]
With API 4.2 support (reading color and brightness settings) and a non-collapsing color-picker, this is close to be merged.

**Comment by bk1285** [comment]
Hey @FrankX0, that was a quick integration of v4.2! I really like the latest version and already announced it here: https://github.com/bk1285/rpi_wordclock#-in-progress :)

**Comment by bk1285** [comment]
Another consideration (.... for a future PR :smiley: ): Currently, the webinterface allows to control Hue and Value (brightness) of the [HSV colorspace](https://en.wikipedia.org/wiki/HSL_and_HSV#/media/File:HSV_color_solid_cone_chroma_gray.png). How about an update of the color ring to a color circle providing the option to control the saturation S as well? What do you think about it?

**Comment by FrankX0** [comment]
I will have a look (.... for a future PR 😃).
Some other thoughts:
- Store the current colour scheme in the config file from the app
- Integration with the auto-brightness functionality (read-only brightness slider, enable/disable auto-brightness)

**Comment by FrankX0** [comment]
Hey Bernd,

Actually, implementing full HSL control was easier than I expected: very good suggestion!

I agree that the number of sliders seems much, but I think we need all these controls.
HSL/HSV controls the characteristics of the colour, but not (directly) the amount of light which is produced. HSL/HSV would be sufficient for reflective colours (printing) but for LEDs (and LCDs) you need this additional control. For example on an LCD screen you can change the colour of a pixel independently from the amount of light which is produced (backlight).

But again: I welcome all improvement ideas!

Mit freundliche grüßen,
Frank.

**Comment by bk1285** [comment]
Hi @FrankX0,

probably it's really more about just setting RGB-values, since the brightness of the LEDs is provided to the LED-strip through an additional parameter... I'll do the merge to develop with the sliders.

Before going to master, I'd like to elaborate on the sliders once again. Meanwhile, I'll add https://github.com/bk1285/rpi_wordclock/pull/100/commits/53c963c536f58fd5ec3958c8b9d91688d666b229 to master.

Best,
Bernd

**Pull request merged**